### PR TITLE
Refactor AST types to be classes with serialize methods.

### PR DESCRIPTION
**Pull request by aomarks**
As suggested by @rictic. What do we think?

Note the diff might be a bit hard to read, but I promise there are no changes to the serialization logic other than refactoring to method style.

I do kind of like the layout of having data structures concisely defined in one file, and all serialization routines defined in another. It organizes the problem space more intuitively to me I think? But the OOP style is more concise overall, avoids the type switches, and is probably more intuitive to other people.

### CODE REVIEW

**Comment by btelles** [comment]
6 of one, half-dozen of the other if you ask me.
As an outsider coming in, either approach is equally cryptic. A sentence or two documenting the overarching idea (what the data structures and classes mean, etc) would go a long way toward helping me grok and contribute to the project.

File: src/ts-ast.ts [hunk 1]
```diff
@@ -9,117 +9,325 @@
...
+  readonly kind = 'class';
+
+  description: string = '';
+  extends: string = '';
+  mixins: string[] = [];
+  properties: Property[] = [];
+  methods: Method[] = [];
+
+  constructor(public name: string) {
+  }
```
Comment (justinfagnani) [inline src/ts-ast.ts]: whither name?

File: src/gen-ts.ts [hunk 2]
```diff
@@ -196,58 +183,44 @@ function handleBehavior(feature: analyzer.PolymerBehavior, root: ts.Document) {
...
-  const parent = findOrCreateNamespace(root, namespacePath);
-  parent.members.push({
-    kind: 'interface',
-    name: className,
-    description: feature.description,
-    extends: [],
-    properties: handleProperties(feature.properties.values()),
-    methods: handleMethods(feature.methods.values()),
-  });
+  const i = new ts.Interface(className);
```
Comment (justinfagnani) [inline src/gen-ts.ts]: I find the declarative style used before a bit easier to read. I if you need classes, you can have a constructor that takes a data object.

**Review by justinfagnani — COMMENTED** [review]
Relying on polymorphism is fine, I could go either way. What I think would help more is decomposing  output into templates and reusing some logic defined in a base class:

```ts
serialize() {
  return unindent`
    class ${name} extends ${this.serializeExtends()} {
      ${this.serializeProperties()};
      ${this.serializeMethods()};
    }
  `;
};
```

File: src/ts-ast.ts (hunk 1 above)
Comment (aomarks) [reply src/ts-ast.ts]: If you annotate a constructor parameter with a visibility keyword, it is automatically assigned as a member variable.

https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties

Too obscure? Second time somebody has asked about it. I do find it makes it harder to scan for "what members does this thing have".

File: src/gen-ts.ts (hunk 2 above)
Comment (aomarks) [reply src/gen-ts.ts]: Using data objects for most classes now.

**Comment by aomarks** [comment]
Added some TODOs for improving documentation and making serialization easier to read.

Gonna merge this for now, as I am getting into a refactoring paralysis.

**Pull request merged**